File: engine/recipe.py
"""
Python Recipe Evaluator — 管道公式版本（Phase 2）

核心理念：金额 = 数量 × 单价
- 数量管道：从 wet_weight 开始，经过零或多步变换
- 单价管道：从输入（固定价/基准价）开始，经过零或多步变换

核心函数：
  evaluate_recipe(recipe, batch_view, direction) -> list[SettlementItemRecord]

设计原则：
  - 通用管道执行器，支持任意变换组合
  - 金额全程 Decimal + ROUND_HALF_UP
  - direction 参数接受 "采购" 或 "销售"
"""
from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

from core.models.batch import Batch, BatchView
from core.models.settlement_item import (
    PricingBasis,
    PriceFormula,
    PriceSource,
    SettlementDirection,
    SettlementItemRecord,
    SettlementRowType,
)

from .schema import PriceStep, QuantityStep, Recipe, TierEntry


# ══════════════════════════════════════════════════════════════
# 数量管道执行器
# ══════════════════════════════════════════════════════════════

def _execute_quantity_pipeline(
    pipeline: list[QuantityStep],
    batch: Batch,
) -> tuple[Decimal, dict]:
    """执行数量管道，返回最终数量和中间值

    Returns:
        (最终数量, 中间值字典) - 中间值包含 dry_weight, metal_quantity 等
    """
    wet_weight = batch.total_wet_weight
    assay = batch.assay_report
    h2o = assay.h2o_pct or Decimal("0")

    # 中间值记录
    intermediates: dict = {}

    # 当前累积值
    current: Decimal | None = None

    for i, step in enumerate(pipeline):
        if step.op == "start":
            if step.operand == "wet_weight":
                current = wet_weight
            elif step.operand == "dry_weight":
                # 单独计算干重并记录
                current = wet_weight * (Decimal("1") - h2o / Decimal("100"))
                intermediates["dry_weight"] = current
            else:
                raise ValueError(f"未知 operand: {step.operand}")

        elif step.op == "dry_weight":
            # 计算干重：wet × (1 - h2o/100)
            current = wet_weight * (Decimal("1") - h2o / Decimal("100"))
            intermediates["dry_weight"] = current

        elif step.op == "multiply_field":
            field = step.field
            if field is None:
                raise ValueError("multiply_field 需要指定 field")

            # 获取字段值
            field_value: Decimal | None = getattr(assay, field, None)
            if field_value is None:
                raise ValueError(f"化验单缺少字段: {field}")

            factor = step.factor if step.factor is not None else Decimal("1")

            if current is None:
                # 如果还没有累积值，直接用字段值
                current = field_value * factor
            else:
                # 乘法：current × field_value × factor
                current = current * field_value * factor

        elif step.op == "grade_adjust":
            """品位计算：计算金属量 = 干重 × 品位

            - unit="pct": 金属量 = 干重 × 品位 ÷ 100
            - unit="gpt": 金属量 = 干重 × 品位（不转换）
            """
            field = step.field
            if field is None:
                raise ValueError("grade_adjust 需要指定 field")

            field_value: Decimal | None = getattr(assay, field, None)
            if field_value is None:
                raise ValueError(f"化验单缺少字段: {field}")

            if current is None:
                raise ValueError("grade_adjust 前需要有累积值（干重）")

            # 根据单位决定是否转换
            if step.unit == "pct":
                # % 单位：除以 100
                current = current * field_value / Decimal("100")
            elif step.unit == "gpt" or step.unit is None:
                # g/t 单位：不转换
                current = current * field_value
            else:
                raise ValueError(f"grade_adjust 不支持的单位: {step.unit}")

            intermediates["metal_quantity"] = current

        elif step.op == "subtract":
            if step.value is None:
                raise ValueError("subtract 需要指定 value")
            if current is None:
                raise ValueError("subtract 前需要有累积值")
            current = current - step.value

        elif step.op == "add":
            if step.value is None:
                raise ValueError("add 需要指定 value")
            if current is None:
                raise ValueError("add 前需要有累积值")
            current = current + step.value

        elif step.op == "multiply":
            if step.value is None:
                raise ValueError("multiply 需要指定 value")
            if current is None:
                raise ValueError("multiply 前需要有累积值")
            current = current * step.value

        elif step.op == "divide":
            if step.value is None or step.value == Decimal("0"):
                raise ValueError("divide 需要指定非零 value")
            if current is None:
                raise ValueError("divide 前需要有累积值")
            current = current / step.value

        else:
            raise ValueError(f"未知数量操作: {step.op}")

    if current is None:
        raise ValueError("数量管道为空")

    return current.quantize(Decimal("0.0001"), rounding=ROUND_HALF_UP), intermediates


# ══════════════════════════════════════════════════════════════
# 单价管道执行器
# ══════════════════════════════════════════════════════════════

def _find_tier(grade: Decimal, tiers: list[TierEntry]) -> TierEntry | None:
    """查找品位对应的阶梯档位

    区间规则：lower 含（闭），upper 不含（开）。
    最高档 upper=None 表示无上限。
    """
    for tier in sorted(tiers, key=lambda t: t.lower):
        if grade < tier.lower:
            continue
        if tier.upper is None:
            return tier
        if grade < tier.upper:
            return tier
    return None


def _execute_price_pipeline(
    pipeline: list[PriceStep],
    batch: Batch,
) -> tuple[Decimal, str | None]:
    """执行单价管道，返回最终单价和字段名

    Returns:
        (最终单价, 品位字段名或None)
    """
    assay = batch.assay_report

    current: Decimal | None = None
    grade_field: str | None = None

    for step in pipeline:
        if step.op == "fixed":
            if step.value is None:
                raise ValueError("fixed 需要指定 value")
            current = step.value

        elif step.op == "multiply":
            if step.factor is None:
                raise ValueError("multiply 需要指定 factor")
            if current is None:
                raise ValueError("multiply 前需要有累积值")
            current = current * step.factor

        elif step.op == "subtract":
            if step.factor is None:
                raise ValueError("subtract 需要指定 factor")
            if current is None:
                raise ValueError("subtract 前需要有累积值")
            current = current - step.factor

        elif step.op == "tier_lookup":
            field = step.field
            if field is None:
                raise ValueError("tier_lookup 需要指定 field")
            if step.tiers is None:
                raise ValueError("tier_lookup 需要指定 tiers")

            grade_field = field

            # 获取字段值（品位）
            grade: Decimal | None = getattr(assay, field, None)
            if grade is None:
                raise ValueError(f"化验单缺少字段: {field}")

            tier = _find_tier(grade, step.tiers)
            if tier is None:
                # 品位未落入任何档位，单价为 0
                current = Decimal("0")
            else:
                current = tier.rate

        else:
            raise ValueError(f"未知单价操作: {step.op}")

    if current is None:
        raise ValueError("单价管道为空")

    return current.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP), grade_field


# ══════════════════════════════════════════════════════════════
# 主函数
# ══════════════════════════════════════════════════════════════

def evaluate_recipe(
    recipe: Recipe,
    batch_view: BatchView,
    direction: str,
) -> list[SettlementItemRecord]:
    """
    用 Recipe（管道公式版本）对 BatchView 中的每个 Batch 计算结算明细。

    Args:
        recipe:      合同计价配方（管道版本）
        batch_view:  批次视图
        direction:   合同方向（"采购" 或 "销售"）

    Returns:
        SettlementItemRecord 列表：
          - 先按 batch 顺序输出各元素货款行
          - 再输出各杂质扣款行（按 recipe.elements 顺序）

    异常：
        ValueError  — 化验单缺少必要字段、步骤参数错误
    """
    settle_direction = (
        SettlementDirection.EXPENSE if direction == "采购"
        else SettlementDirection.INCOME
    )

    items: list[SettlementItemRecord] = []

    element_items = [e for e in recipe.elements if e.type == "element"]
    deduction_items = [e for e in recipe.elements if e.type == "deduction"]

    for batch in batch_view.batches:
        assay = batch.assay_report
        wet_weight = batch.total_wet_weight
        h2o_pct = assay.h2o_pct

        # 验证必要字段
        if h2o_pct is None:
            raise ValueError(
                f"批次 {batch.batch_id} 化验单缺少 h2o_pct，无法计算"
            )

        # ── 计价元素（type="element"）─────────────────────────────
        for elem in element_items:
            # 执行数量管道
            quantity, intermediates = _execute_quantity_pipeline(
                elem.quantity_pipeline, batch
            )

            # 执行单价管道
            unit_price, grade_field = _execute_price_pipeline(
                elem.price_pipeline, batch
            )

            # 计算金额
            amount = (quantity * unit_price).quantize(
                Decimal("0.01"), rounding=ROUND_HALF_UP
            )

            # 获取中间值
            dry_weight_val = intermediates.get("dry_weight")
            metal_quantity_val = intermediates.get("metal_quantity")

            # 金属量四舍五入到3位小数（结算标准）
            if metal_quantity_val is not None:
                metal_quantity_rounded = metal_quantity_val.quantize(
                    Decimal("0.001"), rounding=ROUND_HALF_UP
                )
                amount = (metal_quantity_rounded * unit_price).quantize(
                    Decimal("0.01"), rounding=ROUND_HALF_UP
                )
            else:
                amount = (quantity * unit_price).quantize(
                    Decimal("0.01"), rounding=ROUND_HALF_UP
                )

            # 获取品位值
            grade_value = None
            if grade_field:
                grade_value = getattr(assay, grade_field, None)

            # 构建 SettlementItemRecord
            item = SettlementItemRecord(
                contract_id=recipe.contract_id,
                sample_id=batch.sample_id,
                row_type=SettlementRowType.ELEMENT_PAYMENT,
                direction=settle_direction,
                element=elem.name,
                pricing_basis=PricingBasis.METAL_QUANTITY,
                price_source=PriceSource.FIXED,
                price_formula=PriceFormula.FIXED_PRICE,
                wet_weight=wet_weight,
                h2o_pct=h2o_pct,
                dry_weight=dry_weight_val,
                metal_quantity=metal_quantity_val.quantize(Decimal("0.001"), rounding=ROUND_HALF_UP) if metal_quantity_val else None,
                assay_grade=grade_value,
                unit_price=unit_price,
                unit=elem.unit,
                amount=amount,
            )
            items.append(item)

        # ── 杂质扣款（type="deduction"）──────────────────────────
        for elem in deduction_items:
            # 执行数量管道（通常是干重）
            quantity, intermediates = _execute_quantity_pipeline(
                elem.quantity_pipeline, batch
            )

            # 执行单价管道（阶梯查表）
            unit_price, grade_field = _execute_price_pipeline(
                elem.price_pipeline, batch
            )

            # 如果单价为0（品位不在任何阶梯档位），跳过该扣款记录
            if unit_price == Decimal("0"):
                continue

            # 计算金额（扣款，方向为 EXPENSE）
            amount = (quantity * unit_price).quantize(
                Decimal("0.01"), rounding=ROUND_HALF_UP
            )

            # 获取品位值
            grade_value = None
            if grade_field:
                grade_value = getattr(assay, grade_field, None)

            # 档位说明
            tier_note = None
            for step in elem.price_pipeline:
                if step.op == "tier_lookup" and step.tiers:
                    tier = _find_tier(grade_value or Decimal("0"), step.tiers)
                    if tier:
                        if tier.upper is None:
                            tier_note = f"{elem.name} ≥{tier.lower}% → {tier.rate}元/吨"
                        else:
                            tier_note = f"{elem.name} [{tier.lower}%, {tier.upper}%) → {tier.rate}元/吨"
                    break

            items.append(SettlementItemRecord(
                contract_id=recipe.contract_id,
                sample_id=batch.sample_id,
                row_type=SettlementRowType.IMPURITY_DEDUCTION,
                direction=SettlementDirection.EXPENSE,
                element=elem.name,
                pricing_basis=PricingBasis.WET_WEIGHT,
                price_source=PriceSource.FIXED,
                wet_weight=wet_weight,
                assay_grade=grade_value,
                unit_price=unit_price,
                unit="元/吨",
                amount=amount,
                note=tier_note,
            ))

    return items

File: engine/recipe.js
/**
 * JS Recipe Evaluator — 前端实时预览引擎（Number 精度）
 *
 * 与 engine/recipe.py 逻辑完全一致，但使用 Number 类型（非 Decimal）。
 * 用于前端即时结算预览，正式结算走 Python/Decimal 引擎，容差 0.01 元可接受。
 *
 * 主函数：evaluateRecipe(recipe, batchView, direction) → SettlementItem[]
 *
 * 使用方式（前端）：
 *   import { evaluateRecipe } from '../engine/recipe.js'
 *   const items = evaluateRecipe(recipe, batchView, '采购')
 */

// ── 精度工具 ────────────────────────────────────────────────

/** 四舍五入到 n 位小数 */
function round(value, dp) {
  const factor = Math.pow(10, dp);
  return Math.round(value * factor) / factor;
}

/** 干重 = 湿重 × (1 - H2O% / 100)，保留 4 位小数 */
function calcDryWeight(wetWeight, h2oPct) {
  return round(wetWeight * (1 - h2oPct / 100), 4);
}

/** 金属量 = 干重 × 有效品位 / 100，保留 3 位小数 */
function calcMetalQuantity(dryWeight, assayPct, gradeDeduction) {
  const effGrade = assayPct - gradeDeduction;
  return round(dryWeight * (effGrade / 100), 3);
}

/** 货款 = 金属量 × 单价，保留 2 位小数 */
function calcElementPayment(metalQuantity, unitPrice) {
  return round(metalQuantity * unitPrice, 2);
}

/** 扣款金额 = 湿重 × 费率，保留 2 位小数 */
function calcImpurityAmount(wetWeight, rate) {
  return round(wetWeight * rate, 2);
}

// ── 档位查找 ─────────────────────────────────────────────────

/**
 * 找到 grade 对应的扣款档位。
 * 区间规则：lower 含（闭），upper 不含（开）；upper=null 表示无上限。
 */
function findTier(grade, tiers) {
  const sorted = [...tiers].sort((a, b) => Number(a.lower) - Number(b.lower));
  for (const tier of sorted) {
    const lower = Number(tier.lower);
    if (grade < lower) continue;
    if (tier.upper === null || tier.upper === undefined) return tier;
    const upper = Number(tier.upper);
    if (grade < upper) return tier;
  }
  return null;
}

// ── 主函数 ───────────────────────────────────────────────────

/**
 * 用 Recipe 对 batchView 中的每个批次单元计算结算明细。
 *
 * @param {Object} recipe - Recipe 对象（与 engine/schema.py 的 Recipe 结构一致）
 * @param {Object} batchView - 批次视图 { contract, batchUnits }
 *   batchUnits: Array of { sampleId, totalWetWeight, assayReport }
 *   assayReport: { cuPct, h2oPct, asPct, ... }（驼峰命名）
 * @param {string} direction - "采购" | "销售"
 * @returns {SettlementItem[]}
 */
export function evaluateRecipe(recipe, batchView, direction) {
  const settleDirection = direction === '采购' ? '付' : '收';
  const items = [];

  const elementItems = recipe.elements.filter(e => e.type === 'element');
  const deductionItems = recipe.elements.filter(e => e.type === 'deduction');

  // ── 元素货款 ────────────────────────────────────────────
  for (const unit of batchView.batchUnits) {
    const assay = unit.assayReport;
    const wetWeight = Number(unit.totalWetWeight);
    const h2oPct = Number(assay.h2oPct);

    if (isNaN(h2oPct)) {
      throw new Error(`样号 ${unit.sampleId} 化验单缺少 h2o_pct`);
    }

    const dryWeight = calcDryWeight(wetWeight, h2oPct);

    for (const elem of elementItems) {
      if (elem.unitPrice.source !== 'fixed') {
        throw new Error(`仅支持 source=fixed，元素 ${elem.name} 使用了 ${elem.unitPrice.source}`);
      }
      if (elem.operations && elem.operations.length > 0) {
        throw new Error(`元素 ${elem.name} 含 operations，Phase 1 暂不支持`);
      }

      const unitPrice = Number(elem.unitPrice.value);
      const basis = elem.quantity.basis;

      if (basis === 'wet_weight') {
        const payment = round(wetWeight * unitPrice, 2);
        items.push({
          sampleId: unit.sampleId,
          rowType: '元素货款',
          direction: settleDirection,
          element: elem.name,
          pricingBasis: '湿重',
          wetWeight,
          h2oPct,
          unitPrice,
          unit: elem.unitPrice.unit,
          amount: payment,
        });

      } else if (basis === 'dry_weight') {
        const payment = round(dryWeight * unitPrice, 2);
        items.push({
          sampleId: unit.sampleId,
          rowType: '元素货款',
          direction: settleDirection,
          element: elem.name,
          pricingBasis: '干重',
          wetWeight,
          h2oPct,
          dryWeight,
          unitPrice,
          unit: elem.unitPrice.unit,
          amount: payment,
        });

      } else if (basis === 'metal_quantity') {
        const gradeField = elem.quantity.gradeField;
        if (!gradeField) {
          throw new Error(`元素 ${elem.name} basis=metal_quantity 必须指定 gradeField`);
        }
        const assayGrade = Number(assay[gradeField]);
        if (isNaN(assayGrade)) {
          throw new Error(`样号 ${unit.sampleId} 化验单缺少字段 ${gradeField}`);
        }
        const metalQty = calcMetalQuantity(dryWeight, assayGrade);
        const payment = calcElementPayment(metalQty, unitPrice);

        items.push({
          sampleId: unit.sampleId,
          rowType: '元素货款',
          direction: settleDirection,
          element: elem.name,
          pricingBasis: '金属量',
          wetWeight,
          h2oPct,
          dryWeight,
          assayGrade,
          metalQuantity: metalQty,
          unitPrice,
          unit: elem.unitPrice.unit,
          amount: payment,
        });
      }
    }
  }

  // ── 杂质扣款 ────────────────────────────────────────────
  for (const elem of deductionItems) {
    const gradeField = elem.quantity.gradeField;
    if (!gradeField) {
      throw new Error(`杂质 ${elem.name} 必须指定 gradeField`);
    }

    for (const unit of batchView.batchUnits) {
      const assay = unit.assayReport;
      const gradeRaw = assay[gradeField];
      if (gradeRaw === null || gradeRaw === undefined) continue;

      const grade = Number(gradeRaw);
      const tier = findTier(grade, elem.tiers || []);
      if (!tier) continue;

      const rate = Number(tier.rate);
      const wetWeight = Number(unit.totalWetWeight);
      const amount = calcImpurityAmount(wetWeight, rate);

      const tierNote = tier.upper === null || tier.upper === undefined
        ? `${elem.name} ≥${tier.lower}% → ${tier.rate}元/吨`
        : `${elem.name} [${tier.lower}%, ${tier.upper}%) → ${tier.rate}元/吨`;

      items.push({
        sampleId: unit.sampleId,
        rowType: '杂质扣款',
        direction: '付',   // 杂质扣款永远是支出
        element: elem.name,
        pricingBasis: '湿重',
        wetWeight,
        assayGrade: grade,
        unitPrice: rate,
        unit: '元/吨',
        amount,
        note: tierNote,
      });
    }
  }

  return items;
}

/**
 * 从 SettlementItem[] 计算汇总数据。
 *
 * @param {Object[]} items
 * @returns {{ totalElementPayment, totalImpurityDeduction, totalIncome, totalExpense, netAmount }}
 */
export function summarizeItems(items) {
  let totalElementPayment = 0;
  let totalImpurityDeduction = 0;
  let totalIncome = 0;
  let totalExpense = 0;

  for (const item of items) {
    if (item.rowType === '元素货款') {
      totalElementPayment = round(totalElementPayment + item.amount, 2);
    } else if (item.rowType === '杂质扣款') {
      totalImpurityDeduction = round(totalImpurityDeduction + item.amount, 2);
    }

    if (item.direction === '收') {
      totalIncome = round(totalIncome + item.amount, 2);
    } else {
      totalExpense = round(totalExpense + item.amount, 2);
    }
  }

  const netAmount = round(totalIncome - totalExpense, 2);

  return {
    totalElementPayment,
    totalImpurityDeduction,
    totalIncome,
    totalExpense,
    netAmount,
  };
}

// ── 字段名适配器 ─────────────────────────────────────────────
// API 返回 snake_case，JS 引擎用 camelCase
// 前端从 API 获取数据后需要先经过此函数转换

/**
 * 将 API 返回的 batchView（snake_case）转换为 JS 引擎使用的 camelCase 格式。
 *
 * API batchUnit: { sample_id, total_wet_weight, assay_report: { cu_pct, h2o_pct, ... } }
 * JS engine: { sampleId, totalWetWeight, assayReport: { cuPct, h2oPct, ... } }
 */
export function adaptBatchView(apiBatchView) {
  return {
    contract: apiBatchView.contract,
    batchUnits: apiBatchView.batch_units.map(unit => ({
      sampleId: unit.sample_id,
      totalWetWeight: unit.total_wet_weight,
      assayReport: adaptAssayReport(unit.assay_report),
    })),
  };
}

function adaptAssayReport(ar) {
  return {
    cuPct: ar.cu_pct,
    auGt: ar.au_gt,
    agGt: ar.ag_gt,
    pbPct: ar.pb_pct,
    znPct: ar.zn_pct,
    sPct: ar.s_pct,
    asPct: ar.as_pct,
    h2oPct: ar.h2o_pct,
  };
}

/**
 * 将 Recipe（snake_case，来自 API/YAML）转换为 JS 引擎使用的 camelCase 格式。
 */
export function adaptRecipe(apiRecipe) {
  return {
    contractId: apiRecipe.contract_id,
    version: apiRecipe.version || '1.0',
    elements: (apiRecipe.elements || []).map(elem => ({
      name: elem.name,
      type: elem.type,
      quantity: {
        basis: elem.quantity?.basis || 'metal_quantity',
        gradeField: _snakeToCamel(elem.quantity?.grade_field),
      },
      unitPrice: {
        source: elem.unit_price?.source || 'fixed',
        value: elem.unit_price?.value,
        unit: elem.unit_price?.unit || '元/金属吨',
      },
      operations: elem.operations || [],
      tiers: (elem.tiers || []).map(t => ({
        lower: t.lower,
        upper: t.upper,
        rate: t.rate,
      })),
    })),
    assayFee: apiRecipe.assay_fee,
  };
}

/** grade_field 从 snake_case 转 camelCase（cu_pct → cuPct）*/
function _snakeToCamel(s) {
  if (!s) return s;
  return s.replace(/_([a-z])/g, (_, c) => c.toUpperCase());
}
